### ISSUE TITLE
feedback for scpub3 standalone HTML

### ISSUE THREAD

**Issue by sujato** [suttacentral/publications#56]
This is the file I am commenting on:

[scpub3.zip](https://github.com/suttacentral/publications/files/8283504/scpub3.zip)

and this is the relevant config file:

https://github.com/suttacentral/sc-data/blob/master/sc_bilara_data/_publication/en/sujato/editions/mn/mn-en-sujato_scpub3-ed6-html_2022-02-10.json

What I'll try to do is look at all the issues I come across and specify them with reference to the original spec where relevant.

## foreword

ok! Except we have 

```
</article></body></html>
<html><body><article
```

It should be

```
</article><article
```
- [x] Validate the HTML! I ran 

```
tidy --doctype html5 --output-html 1 --tidy-mark 0 --quiet 1 --output-encoding utf8 -w 0 --show-warnings 0 -m scpub3.html
```

This created the following file, which I will use from now on.

[scpub3-tidy.zip](https://github.com/suttacentral/publications/files/8283676/scpub3-tidy.zip)

## introduction

ok!

## missing things

Front matter is incomplete, according to editions config we must have:

- [x]  "titlepage",
- [x]  "copyright",
- [x]  "./matter/epigraph.html",
- [x]  "toc",
- [X]  "./matter/foreword.html",
- [x]  "./matter/preface.html",
- [X]  "./matter/introduction.html",
- [X]  "./matter/acknowledgements.html",
- [x]  "blurbs"

If any of these are underspecified in the project issues, let me know. I will review the issues to make sure they are all complete.

## html bugs

At mn12:50.7 there is an unclosed blockquote. In fact the verse is missing an entire segment, mn12:50.10

see source file here:

https://github.com/suttacentral/sc-data/blob/master/sc_bilara_data/html/pli/ms/sutta/mn/mn12_html.json

There's also content missing a little above that, at mn12:50.2

Actually when I look, there is missing content all over the place. Clearly there's some major bug in getting the text into the file.

Here's the first sutta, with the HTML standalone next to the same sutta as it appears on our site. After the first couple of paragraphs, all the remaining visible text is missing.

![Screenshot from 2022-03-17 21-14-27](https://user-images.githubusercontent.com/6112010/158787793-24d3bdc9-dd1f-49f2-bd5f-75ce0031bdb9.png)

The text is also disordered. Here you can see highlighted the last line of the Sutta.

![Screenshot from 2022-03-17 21-17-20](https://user-images.githubusercontent.com/6112010/158788095-1a753d60-f9ce-4a40-adc8-272b1378eb28.png)

But then it jumps back to earlier, and the `<article>` is closed improperly.

Okay, so there's a bunch of bugs here, I won't keep reporting them all!

## reference markup

-  **I am changing the reference spec from what I said before!**

Here is the new spec. But wait on this, I may make more changes.

https://github.com/suttacentral/suttacentral/issues/2417

Take out the `<span id` tags and the reference tags (`<a class="bj"`, `<a class="pts-vp-pli"`)

## space bugs

- [x] Another problem here, is that the sentences don't always have a space at the end.

```
in the forest by day.But in the last month of summer
```

As far as I know, the bilara text you are consuming should be normalized, so that every segment ends in a space, except when the last character is em-dash.

On the website the same sentence is fine:

https://suttacentral.net/mn12/en/sujato

```
in the forest by day. But in the last month of summer 
```

## no html entities

- [x] don't include any HTML entities

```
&amp;
```

Don’t let any of our transformers change anything in the content. Everything remains the same Unicode entity all the way through. When encoding and re-encoding back and forth there are *always* bugs, and there is no need for it. Just force everything to use Unicode all the time.

Same thing applies, even more so, in LaTeX. There are a couple of control characters that need to be escaped, but basically everything should be unicode all the time.

## file name

- [x] Make sure the outputted file has the correctly defined name. `scpub3.html` is ambiguous. Use the same name as the JSON config file:

```
mn-en-sujato_scpub3-ed6-html_2022-02-10.html
```

We must always keep an unambiguous, one-to-one relation between the config and the produced file. As the config says, so the file is produced.

**Comment by sujato** [comment]
Also, we need to supply structural markup for the whole thing:

```
<main>
    <section id='frontmatter'>…</section>
    <section id='mainmatter'>…</section>
    <section id='backmatter'>…</section>
</main>
```

See the demo page for this.

https://github.com/suttacentral/suttacentral/issues/2365

**Comment by sujato** [comment]
The links are not formed properly, once again this is defined in the demo:



We also have reference links. Let's make them work!

    href="/dn ---> href="#dn
    href="/ ---> href="https://suttacentral.net/

Now all the internal links will work in the file, the external links will point to SC.

**Comment by sujato** [comment]
CSS needs to be added:

https://github.com/suttacentral/suttacentral/issues/2412

**Comment by sujato** [comment]
I have tested the text content in scpub3 according to the latest version sent. Here's the source file I used:

[scpub3.zip](https://github.com/suttacentral/publications/files/8322055/scpub3.zip)

To compare, I extracted the contents of `mn` using bilara i/o:

```
./sheet_export.py mn mn.tsv
```

I copied out the translation column, removed the column header, and removed `*` markdown.

Next I opened scpub3.html in a browser, copied it and pasted as plain text. Now I have two plain text files.

Then I removed all new lines and spaces.

```
tr -d '\n' < mn-st.txt > mn-st2.txt
tr -d ' ' < mn-st2.txt > mn-st3.txt
```

Then I divided it into lines of exactly 100 characters:

```
sed -i mn-io3.txt -e "s/.\{100\}/&\n/g"
```

And opened both files in Meld. They're identical!

So i can confirm that in the case of scpub3 there is no text loss.
